src/charmlibs/pathops/_types.py

```diff
@@ -34,138 +34,450 @@
 # based on typeshed.stdlib.pathlib.PurePath
 # https://github.com/python/typeshed/blob/main/stdlib/pathlib.pyi#L29
 class PathProtocol(typing.Protocol):
+    """The protocol implemented by :class:`ContainerPath` and :class:`LocalPath`.
+
+    .. tip::
+        Use this class in type annotations where either :class:`ContainerPath` or
+        :class:`LocalPath` is acceptable. This will result in both correct type checking
+        and useful autocompletions.
+
+        Using a union type instead will also give correct type checking
+        results, but will autocomplete methods and attributes that *any* of the union members have,
+        rather than only those that *all* of the union members have.
+
+        Consider using the following pattern if you don't want callers to have to wrap their
+        arguments with :class:`LocalPath`, but want to use PathProtocol internally::
+
+            def fn(arg: str | os.PathLike[str] | ContainerPath):
+                path: PathProtocol = (
+                    arg if isinstance(arg, ContainerPath) else LocalPath(arg)
+                )
+    """
+
     #############################
     # protocol PurePath methods #
     #############################
 
-    def __hash__(self) -> int: ...
+    def __hash__(self) -> int:
+        """Protocol implementers are hashable."""
+        ...
 
     # comparison methods
-    def __lt__(self, other: Self) -> bool: ...
-    def __le__(self, other: Self) -> bool: ...
-    def __gt__(self, other: Self) -> bool: ...
-    def __ge__(self, other: Self) -> bool: ...
-    def __eq__(self, other: object, /) -> bool: ...
+    def __lt__(self, other: Self) -> bool:
+        """Comparison methods other than equality are only guaranteed for members of the same type.
+
+        Specifically, :class:`ContainerPath` objects are not comparable to other types of objects.
```

Comment:
Similar here, I think we should probably just drop this. Or mention it once on the first comparison method or at the class level.